File: Kaustubh_19gcebit088/Day1.java
// https://leetcode.com/problems/backspace-string-compare/

// time complexity = O(n)
// space complexity = O(1)

// approach:
// traverse through the whole char arr 
// for every element which is not '#'
//     add it to ans string
// and for every element which is '#', 
//     remove the last added element from ans string only if it is not empty



public class Day1<ans> {
    public static void main(String[] args) {
        String a = "ab#c";
        String b = "ab#c";
        System.out.println(getStr(a));
    }

    static String getStr(String a) {
        char[] ch = (a).toCharArray();
        String ans = "";
        int len = ch.length - 1;
        int i = 0;
        while (i <= len) {
            if (ch[i] != '#') {
                ans += ch[i];

            } else if (ans.length() != 0) {
                ans = ans.substring(0, ans.length() - 1);
            }
            ++i;

        }
        return ans;
    }
}

File: Aman Dwivedi 18-ECE/Day 1.cpp
https://leetcode.com/problems/backspace-string-compare/
Time: O(N)
Space: O(N)

class Solution {
public:
    
    string solve(string str)
    {
        int l=str.length();
        stack<int> st;
        for(int i=0;i<l;i++)
        {
            if(str[i]!='#')
                st.push(str[i]);
            else if(str[i]=='#'&&!st.empty())
                st.pop();
        }
        string ans="";
        while(!st.empty())
        {
            char ch=st.top();
            ans+=ch;
            st.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
    bool backspaceCompare(string s, string t) {
        
        string s1=solve(s);
        string s2=solve(t);
        
        if(s1==s2)
            return true;
        else
            return false;
        
    }
};
